neo/rawio/medrawio.py

```diff
@@ -0,0 +1,358 @@
+    with the beginning of the session being timestamp zero.  If UTC timestamps
+    are desired, then the keep_original_times flag in the constructor can be
+    set to True (it defaults to False) and the timestamps used in the object
+    will be seconds, reference to midnight 1 Jan 1970 (assuming that that
+    data is available in the MED data session).
+    """
+
+    extensions = ['medd', 'rdat', 'ridx']
+    rawmode = 'one-dir'
+
+    def __init__(self, dirname=None, password=None, keep_original_times=False, **kargs):
+        BaseRawIO.__init__(self, **kargs)
+        
+        import dhn_med_py
+        from dhn_med_py import MedSession
+    
+        self.dirname = str(dirname)
+        self.password = password
+        self.keep_original_times = keep_original_times
+
+    def _source_name(self):
+        return self.dirname
+
+    def _parse_header(self):
+
+        import dhn_med_py
+        from dhn_med_py import MedSession
+        
+        # Set a default password to improve compatibility and ease-of-use.
+        # This password will be ignored if an unencrypted MED dataset is being used.
+        if self.password is None:
+            self.password = "L2_password"
+            
+        # Open the MED session (open data file pointers and read metadata files)
+        self.sess = MedSession(self.dirname, self.password)
+        
+        # set the matrix calls to be "sample_major" as opposed to "channel_major"
+        self.sess.set_major_dimension("sample")
+        
+        # find number of segments
+        sess_contigua = self.sess.session_info['contigua']
+        self._nb_segment = len(sess_contigua)
+        
+        # find overall session start time.
+        self._session_start_time = sess_contigua[0]['start_time']
+        
+        # keep track of offset from metadata, if we are keeping original times.
+        if not self.keep_original_times:
+            self._session_time_offset = 0 - self._session_start_time
+        else:
+            self._session_time_offset = self.sess.session_info['metadata']['recording_time_offset']
+        
+        # find start/stop times of each segment
+        self._seg_t_starts = []
+        self._seg_t_stops = []
+        for seg_idx in range(self._nb_segment):
+            self._seg_t_starts.append(sess_contigua[seg_idx]['start_time'])
+            self._seg_t_stops.append(sess_contigua[seg_idx]['end_time'])
+        
+        # find number of streams per segment
+        self._stream_info = []
+        self._num_stream_info = 0
+        for chan_idx, chan_info in enumerate(self.sess.session_info['channels']):
+            chan_freq = chan_info['metadata']['sampling_frequency']
+            
+            # see if we need a new stream, or add channel to existing stream
+            add_to_existing_stream_info = False
+            for stream_info in self._stream_info:
+                if chan_freq == stream_info['sampling_frequency']:
```

Comment:
We'll have to look into this one a little more.  There's not currently a mechanism to simply read the total number of samples from a channel during the session open phase.  We could add this feature.  I just mentioned it to my colleague Matt Stead, and we think that with all the current MED acquisition software in use now, there will always be the same number of samples for channels with the same sampling frequency.  It is true that potential future software written by people other than Dark Horse Neuro might not make that assumption.  We'll think about this one a bit.

Reply:
Thinking a bit more about this one: we already calculate the number of samples in a stream using a "reference" channel (which is the first channel we find to be in a stream).  We could just do that same calculation of every channel prior to making the determination of whether it should be put into an existing stream.  I do worry it might find cases of a single sample at the end of a long stream not being present, due to manner in which the acquisition was stopped, causing the entire channel to be put in a new stream.  But I don't know of any actual cases where this has happened.